[Rezwan Arefin] iP

RezwanArefin01 opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

All you need to do is,
1. download it from [here](https://github.com/RezwanArefin01/ip/releases).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you. :wink: 

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

--------------------------------------------

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

RezwanArefin01 committed:
Level 1 (Greet, Echo, Exit) complete

RezwanArefin01 committed:
Add functionality to add a task and show them

RezwanArefin01 committed:
Separate driver program to Main.java

RezwanArefin01 committed:
Move logic of each command to its own function

RezwanArefin01 committed:
Add the Task class

RezwanArefin01 committed:
Make Duke use the Task class

RezwanArefin01 committed:
Added ability to check and uncheck tasks

RezwanArefin01 committed:
Seperate I/O Operations in IOHelper class.

RezwanArefin01 committed:
Make Task class package private

RezwanArefin01 committed:
Move command implementations in Command class

RezwanArefin01 committed:
Support different types of tasks: ToDo, Deadline, Event

RezwanArefin01 committed:
Strip around arguments; allow pressing Enter without any commands

RezwanArefin01 committed:
Fix index out of range bug in update status command

RezwanArefin01 committed:
Add text based test cases

RezwanArefin01 committed:
Add DukeException class; handle exceptions using it

RezwanArefin01 committed:
Added delete command; updated tests.

RezwanArefin01 committed:
Add ACTUAL.TXT to .gitignore

RezwanArefin01 committed:
Add the TaskList class as a replacement for ArrayList<Task>

RezwanArefin01 committed:
Replace all ArrayList<Task> with TaskList

RezwanArefin01 committed:
Use LocalDateTime in Task class instead of String

RezwanArefin01 committed:
Merge branch 'branch-Level-7'

RezwanArefin01 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Task.java

RezwanArefin01 committed:
Fix bug: Task::toSaveString() returns time in default format instead of input format

RezwanArefin01 committed:
Seperate UI related methods in UI class, previously IOHelper

RezwanArefin01 committed:
Organize codebase into packages: duke, duke.task, duke.command, duke.util

RezwanArefin01 committed:
Move inner classes of Command and Task outside

RezwanArefin01 committed:
Move sanity check of argument list to Command constructor and rename runSpecialTask to execute

RezwanArefin01 committed:
Separate storage and parsing elements in Storage and Parser class respectively

RezwanArefin01 committed:
Add TaskSymbolType Enum instead of hardcoded symbols

RezwanArefin01 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

RezwanArefin01 committed:
Configure project to use Gradle

RezwanArefin01 committed:
Move date-time format to Parser and fix crashing when data file is corrupted

RezwanArefin01 committed:
Upgrade JUnit to 5.9.0

RezwanArefin01 committed:
Handle parsing exceptions with ParseException

RezwanArefin01 committed:
Add Parser test cases

RezwanArefin01 committed:
Add TaskList test cases

RezwanArefin01 committed:
Make TaskList Iterable

RezwanArefin01 committed:
Use for-each style loop for TaskList

RezwanArefin01 committed:
Add FindCommand

RezwanArefin01 committed:
Add Parser test cases for find command

RezwanArefin01 committed:
Add JavaDocs

RezwanArefin01 committed:
Add Checkstyle support

RezwanArefin01 committed:
Modify test codes to comply with coding standard

RezwanArefin01 committed:
Modify main codes to comply with coding standard

RezwanArefin01 committed:
Organize imports

RezwanArefin01 committed:
Merge branch 'branch-A-JavaDoc'

RezwanArefin01 committed:
Merge branch 'A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/task/DeadlineTask.java
#	src/main/java/duke/task/EventTask.java
#	src/main/java/duke/task/TaskList.java
#	src/main/java/duke/util/Parser.java
#	src/main/java/duke/util/Storage.java

RezwanArefin01 committed:
Merge branch 'branch-Level-9'

RezwanArefin01 committed:
Add JavaDocs to FindCommand

KSHan29 commented on src/main/java/duke/command/CommandType.java:
```diff
@@ -0,0 +1,32 @@
+     * @param args The arguments to check.
+     * @return {@code true} if the arguments are compatible with the command, {@code false} otherwise.
+     */
+    public boolean isCompatible(String[] args) {
```
Perhaps could use verbs for naming so it represents the action of checking compatibility.

RezwanArefin01 replied:
I have named it this way as it returns a boolean. It sounds correct when used: `if (type.isCompatible(args))`. So, the `is` is the verb here.

KSHan29 commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,94 @@
+    /**
+     * Ensures that the {@code dataFile} is there and accessible.
+     */
+    private void ensureDataFile() {
```
I like that the JavaDoc comment describes the comment well. But perhaps you could add 'Exists' to the end of the name for method name to be more descriptive.

KSHan29 commented on src/main/java/duke/util/UI.java:
```diff
@@ -0,0 +1,85 @@
+/**
+ * A class that provides a simple user interface for the Duke system.
+ */
+public class UI {
```
Since UI is an abbreviation, perhaps this could be named Ui instead to follow the recommended coding guidelines

RezwanArefin01 replied:
Refactored the code to use `Ui`. I missed this rule from the coding guidelines.

KSHan29 commented on src/main/java/duke/task/TaskSymbolType.java:
```diff
@@ -0,0 +1,25 @@
+    T("\uD83D\uDCDD"),
+    D("⏰"),
+    E("\uD83D\uDCC5");
+
```
I like that you added this!

KSHan29 reviewed (commented):
Overall lgtm!
I like that you added some cool modifications such as adding logos.
Some methods could be named more descriptively but they are good as it is too.

RezwanArefin01 committed:
Update code to follow coding guidelines

RezwanArefin01 committed:
Merge branch 'branch-A-CheckStyle'

RezwanArefin01 committed:
Add a construction with vaargs for TaskList

RezwanArefin01 committed:
Use varargs for TaskList testing

RezwanArefin01 committed:
Make CommandType::isCompatible use varargs

RezwanArefin01 committed:
Merge branch 'branch-A-Varargs'

EiffelLKF commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,154 @@
+ */
+public class Parser {
+
+    public static final DateTimeFormatter DATE_TIME_INPUT_FORMAT = DateTimeFormatter.ofPattern("d-M-yy HHmm");
+    public static final DateTimeFormatter DATE_TIME_OUTPUT_FORMAT = DateTimeFormatter.ofPattern("hh:mm a, MMM d, yyyy");
```
You might want to add header comments to explain what these fields are used for. I think this applies to the fields in the other classes as well, especially the public ones, even though you already did a great job in the naming of each field to be quite self-evident

EiffelLKF commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,94 @@
+     *
+     * @param tasks The {@code TaskList} to save.
+     */
+    public void save(TaskList tasks) {
+        ensureDataFileExists();
+        try {
+            FileWriter fw = new FileWriter(dataFile);
+            for (Task task : tasks) {
+                fw.write(task.toSaveString() + System.lineSeparator());
+            }
+            fw.close();
+        } catch (IOException e) {
+            throw new DukeException("Could not write to file: " + dataFile);
+        }
+    }
+}
```
You may want to include 'throws DukeException' as part of your method definition for your methods that can throw this exception.

EiffelLKF commented on src/test/java/duke/task/TaskListTest.java:
```diff
@@ -0,0 +1,100 @@
+
+import duke.DukeException;
+
+class TaskListTest {
```
You can consider adding class and method header comments for your various test classes to help developers understand what exactly you are testing for in each test case / test class

EiffelLKF commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,11 @@
+ */
+public class DukeException extends RuntimeException {
+
+    public DukeException(String message) {
```
You might need a method header comment here.

EiffelLKF reviewed (commented):
Overall, your code looks really good and clean! I learnt alot from reading how to went about designing your application! Other than a few small suggestions, I think you did a great job! 😃

RezwanArefin01 committed:
Add JavaFX support

RezwanArefin01 committed:
Add GUI printing hello world, remove text-ui-test

RezwanArefin01 committed:
Make Commands return String

RezwanArefin01 committed:
Add minimal working GUI for Duke

RezwanArefin01 committed:
Use Consumer<String> for printing purposes

RezwanArefin01 committed:
Fix ExitCommand

RezwanArefin01 committed:
Fix DialogBox Label getting cutoff vertically

RezwanArefin01 committed:
Add functionality to scroll with mouse

RezwanArefin01 committed:
Fix JavaFX version warnings

RezwanArefin01 committed:
Add chat bubbles

RezwanArefin01 committed:
Fix DialogBox height getting shorter

RezwanArefin01 committed:
Add JavaDocs for GUI classes

RezwanArefin01 committed:
Merge branch 'branch-Level-10'

RezwanArefin01 committed:
Add test for parseTask with less arguments

RezwanArefin01 committed:
Set up Github Actions Continuous Integration

RezwanArefin01 committed:
Use assertions in TaskList get and remove methods

The index verification is done by TaskList's get and remove methods.
This creates a confusion when reading the code in UpdateStatusCommand
and DeleteTaskCommand: why aren't we throwing an Exception if index is
not valid?

Having the index checking to Command's side would improve readability.

Let's move the index checking part to the commands and use assertions
in the TaskList instead.

RezwanArefin01 committed:
Rewrite TaskList#toString() concisely using StringJoiner

RezwanArefin01 committed:
CommandType::isCompatible: Assert that args is non null

RezwanArefin01 committed:
Use Stream to write FindCommand::execute concisely

RezwanArefin01 committed:
Make Task::contains case insensitive

RezwanArefin01 committed:
Refactor Parser to improve Code Quality.

The Parser::parseCommand method is not compliant with the SLAP.

Let's move the token extraction segment and extracting command type
(with possibly throwing exception) segment to their own methods. Let's
also include the command itself inside argument list, since this is
the general way how argument lists are represented.

RezwanArefin01 committed:
Refactor Storage::load() for improved Code Quality

RezwanArefin01 committed:
Make Parser::parseTask not accept any other isDone value than 0 or 1

RezwanArefin01 committed:
Refactor Task::parseTask to improve Code Quality

RezwanArefin01 committed:
Add minimal help command

RezwanArefin01 committed:
Merge pull request #1 from RezwanArefin01/branch-A-Assertions

Branch A-Assertions

RezwanArefin01 committed:
Merge pull request #2 from RezwanArefin01/branch-A-Streams

Branch A-Streams

RezwanArefin01 committed:
Merge branch 'master' into branch-A-CodeQuality

RezwanArefin01 committed:
Merge pull request #3 from RezwanArefin01/branch-A-CodeQuality

Branch A-CodeQuality

RezwanArefin01 committed:
Remove TaskList index out of bound tests due to it being verified by assertions

RezwanArefin01 committed:
Reduce magic constants in task and command packages

RezwanArefin01 committed:
Split util package into parser and storage packages

RezwanArefin01 committed:
Give more meaningful error messages

RezwanArefin01 committed:
Show error if file permissions are not correct

RezwanArefin01 committed:
Add title and icon

RezwanArefin01 committed:
Add sample UI screenshot

RezwanArefin01 committed:
Allow multiple datetime formats

RezwanArefin01 committed:
Add user guide draft

RezwanArefin01 committed:
Add jekyll theme

RezwanArefin01 committed:
Change jekyll theme to minimal

RezwanArefin01 committed:
Update user guide theme

RezwanArefin01 committed:
Update user guide style

RezwanArefin01 committed:
Enable assertions

RezwanArefin01 committed:
Add javafx dependencies for mac and windows

RezwanArefin01 committed:
Improve User Guide

RezwanArefin01 committed:
Change data directory to support MAC

RezwanArefin01 committed:
Remove space from nested ul tags

RezwanArefin01 committed:
Restore README.md

RezwanArefin01 committed:
Add note to not use double-click

RezwanArefin01 committed:
Include SourceCodePro font

RezwanArefin01 committed:
Add note to install source code pro font
